vsintegration/src/FSharp.Editor/CodeFix/ReplaceWithSuggestion.fs

```diff
@@ -2,45 +2,62 @@
 
-open System
 open System.Composition
-open System.Collections.Immutable
 open System.Threading.Tasks
 
-open Microsoft.CodeAnalysis
 open Microsoft.CodeAnalysis.Text
 open Microsoft.CodeAnalysis.CodeFixes
-open SymbolHelpers
-open Microsoft.FSharp.Compiler.SourceCodeServices.Keywords
+
+open Microsoft.FSharp.Compiler
+open Microsoft.FSharp.Compiler.SourceCodeServices
 
 [<ExportCodeFixProvider(FSharpConstants.FSharpLanguageName, Name = "ReplaceWithSuggestion"); Shared>]
-type internal FSharpReplaceWithSuggestionCodeFixProvider() =
+type internal FSharpReplaceWithSuggestionCodeFixProvider
+    [<ImportingConstructor>]
+    (
+        checkerProvider: FSharpCheckerProvider, 
+        projectInfoManager: FSharpProjectOptionsManager,
+        settings: EditorOptions
+    ) =
     inherit CodeFixProvider()
+
+    static let userOpName = "ReplaceWithSuggestionCodeFix"
     let fixableDiagnosticIds = set ["FS0039"; "FS1129"; "FS0495"]
-    let maybeString = FSComp.SR.undefinedNameSuggestionsIntro()
+    let checker = checkerProvider.Checker
         
     override __.FixableDiagnosticIds = Seq.toImmutableArray fixableDiagnosticIds
 
     override __.RegisterCodeFixesAsync context : Task =
-        async { 
-            context.Diagnostics 
-            |> Seq.filter (fun x -> fixableDiagnosticIds |> Set.contains x.Id)
-            |> Seq.iter (fun diagnostic ->
-                let message = diagnostic.GetMessage()
-                let parts = message.Split([| maybeString |], StringSplitOptions.None)
-                if parts.Length > 1 then
-                    let suggestions = 
-                        parts.[1].Split([| '\r'; '\n' |], StringSplitOptions.RemoveEmptyEntries) 
-                        |> Array.map (fun s -> s.Trim())
-                    
-                    let diagnostics = ImmutableArray.Create diagnostic
-
-                    for suggestion in suggestions do
-                        let replacement = QuoteIdentifierIfNeeded suggestion
-                        let codefix = 
-                            createTextChangeCodeFix(
-                                FSComp.SR.replaceWithSuggestion suggestion, 
-                                context,
-                                (fun () -> asyncMaybe.Return [| TextChange(context.Span, replacement) |]))
-                        context.RegisterCodeFix(codefix, diagnostics))
-        } |> RoslynHelpers.StartAsyncUnitAsTask(context.CancellationToken)
+        asyncMaybe {
+            do! Option.guard settings.CodeFixes.SuggestNamesForErrors
+
+            let document = context.Document
+            let! _, projectOptions = projectInfoManager.TryGetOptionsForEditingDocumentOrProject(document, context.CancellationToken)
+            let! _, _, checkFileResults = checker.ParseAndCheckDocument(document, projectOptions, userOpName=userOpName)
+            
+            let! allSymbolUses = checkFileResults.GetAllUsesOfAllSymbolsInFile() |> liftAsync
```

Comment:
Wouldn't it be better to get the list of intellisense completions for the context? Should be fast enough.